File: W9/class-problems/imposters.py
from collections import deque

def main():
    n, k = map(int, input().strip().split())
    adj = [[] for _ in range(n)]

    for i in range(n):
        data = list(map(int, input().split()))
        s = data[0]
        for u in data[1:]:
            u -= 1
            adj[u].append(i)  # edge (u -> i) in reversed graph

    # run BFS from every villager
    # BE CAREFUL, THE INDENTATION REALLY MATTERS THIS TIME
    for v in range(n):
        # run BFS from v
        visited = [False] * n
        q = deque()

        # initialize BFS: q contains the current wave frontier
        q.append(v)
        visited[v] = True

        while q:
            cur = q.popleft()
            for neighbor in adj[cur]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    q.append(neighbor)

        # compute how many were visited
        count = 0
        for i in range(n):
            if visited[i]:
                count += 1

        if count > k:
            print(1)
        else:
            print(0)

if __name__ == '__main__':
    main()

File: W9/class-problems/imposters.cpp
// kattis problem: https://open.kattis.com/problems/investigatingimposters

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <stack>
#include <map>
using namespace std;

int main() {
    int n,k;
    cin >> n >> k;
    vector<vector<int>> adj(n, vector<int>());

    for(int i = 0; i < n; i++) {
        int s;
        //vilager i
        cin >> s; // size of the list of villager i
        for(int j = 0; j < s; j++) {
            int u;
            cin >> u;
            u--;
            adj[u].push_back(i); // add edge (u,i) in G^R
        }
    }

    // run BFS from every villager
    for(int v = 0; v < n; v++) {
        // run BFS from v
        vector<bool> visited(n, false);
        queue<int> q;
        //initialize BFS: q contains the current wave frontier
        q.push(v);
        visited[v] = true;
        while(!q.empty()) {
            int cur = q.front();
            q.pop();
            for(int neighbor : adj[cur]) {
                if(!visited[neighbor]) {
                    visited[neighbor] = true;
                    q.push(neighbor);
                }
            }
        }
        // compute how many were visited
        int count = 0;
        for(int i = 0; i < n; i++) {
            if(visited[i])
                count++;
        }
        if(count > k)
            cout << 1 << endl;
        else
            cout << 0 << endl;
    }
    return 0;
}